internal/pkg/vm/internal/vmcontext/vmcontex2.go

```diff
@@ -0,0 +1,543 @@
+
+			if r := recover(); r == nil {
+				switch r.(type) {
+				case runtime.AbortPanicError:
+					// Review: abort has a msg inside.. should we log it or what?
+					out = message.Failure(exitcode.MethodAbort).WithGas(gasTank.GasConsumed())
+					return
+				case exitcode.ExitCode:
+					out = message.Failure(r.(exitcode.ExitCode)).WithGas(gasTank.GasConsumed())
+					return
+				default:
+					// Review: how do we rethrow while maintaining call stack?
+					panic(r)
+				}
+			}
+		}()
+
+		// invoke the msg
+		ret := ctx.invoke()
+
+		// set return value
+		out = message.Value(ret).WithGas(gasTank.GasConsumed())
+
+		return
+	}
+	receipt := safeProcess()
+
+	// post-send
+	// 1. charge gas for putting the return value on the chain
+	// 2. success!
+
+	// 1. charge for the space used by the return value
+	// Note: the GasUsed in the message receipt does not
+	ok = gasTank.TryCharge(gascost.OnChainReturnValue(&receipt))
+	if !ok {
+		// Insufficient gas remaining to cover the on-chain return value; proceed as in the case
+		// of method execution failure.
+
+		// get miner owner
+		minerOwner := vm.getMinerOwner(miner)
+
+		// have the sender pay the outstanding bill and get the rest of the money back
+		vm.settleGasBill(msg.From, &gasTank, minerOwner, msg.GasPrice)
+
+		// Note: we are charging the caller not the miner, there is ZERO miner penalty
+		return message.Failure(exitcode.OutOfGas).WithGas(gasTank.GasConsumed()), types.ZeroAttoFIL
+	}
+
+	// 2. Success!
+	return receipt, types.ZeroAttoFIL
+}
+
+func (vm *VM) getMinerOwner(minerAddr address.Address) address.Address {
+	// resolve the target address via the InitActor, and attempt to load state.
+	minerActorEntry, err := vm.state.GetActor(context.Background(), minerAddr)
+	if err != nil {
+		panic("unreachable")
+	}
+
+	// build state handle
+	var stateHandle = NewReadonlyStateHandle(vm.LegacyStorage(), minerActorEntry.Head)
+
+	// get a view into the actor state
+	minerView := miner.NewView(stateHandle, vm.LegacyStorage())
+
+	return minerView.Owner()
+}
+
+func (vm *VM) settleGasBill(sender address.Address, gasTank *gas.Tracker, payee address.Address, gasPrice types.AttoFIL) {
+	// release unused funds that were withheld
+	vm.transfer(address.BurntFundsAddress, sender, gasTank.RemainingGas().Cost(gasPrice))
+	// pay miner for gas
+	vm.transfer(address.BurntFundsAddress, payee, gasTank.GasConsumed().Cost(gasPrice))
+}
+
+// transfer debits money from one account and credits it to another.
+//
+// WARNING: this method will panic if the the amount is negative, accounts dont exist, or have inssuficient funds.
+//
+// Note: this is not idiomatic, it follows the Spec expectations for this method.
+func (vm *VM) transfer(debitFrom address.Address, creditTo address.Address, amount types.AttoFIL) {
+	// allow only for positive amounts
+	if amount.IsNegative() {
+		panic("unreachable: negative funds transfer not allowed")
+	}
+
+	ctx := context.Background()
```

Comment:
Please thread the necessary context here.

If the VM were constructed as a use-once object, it would probably be ok to put the context as a field on vm.